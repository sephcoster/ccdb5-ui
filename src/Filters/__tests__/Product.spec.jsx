import React from 'react'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux'
import { shallow } from 'enzyme';
import ReduxProduct, {Product, mapStateToProps} from '../Product'
import { slugify } from '../../utils'

const fixture = [
  {
    "sub_product.raw": {
      "buckets": [
        {"key": "Credit reporting","doc_count": 3200},
        {"key": "Other personal consumer report","doc_count": 67},
        {"key": "Credit repair services","doc_count": 10}
      ],
    },
    "key": "Credit reporting, credit repair services, or other personal consumer reports",
    "doc_count": 3277
  },
  {
    "sub_product.raw": {
      "buckets": [
        {"key": "Conventional home mortgage","doc_count": 652},
        {"key": "Conventional fixed mortgage","doc_count": 612}
      ],
    },
    "key": "Mortgage",
    "doc_count": 2299
  },
  {
    "sub_product.raw": {
      "buckets": [],
    },
    "key": "Credit reporting",
    "doc_count": 1052
  },
  {
    "sub_product.raw": {
      "buckets": [],
    },
    "key": "Student loan",
    "doc_count": 959
  },
  {
    "sub_product.raw": {
      "buckets": [],
    },
    "key": "Credit card or prepaid card",
    "doc_count": 836
  },
  {
    "sub_product.raw": {
      "buckets": [],
    },
    "key": "Credit card",
    "doc_count": 652
  }
]

function setupEnzyme(initial) {
  const props = {
    options: initial
  }

  const target = shallow(<Product {...props} />);

  return {
    props,
    target
  }
}

function setupSnapshot(initial) {
  const middlewares = [thunk]
  const mockStore = configureMockStore(middlewares)
  const store = mockStore({
    query: {},
    aggs: {
      product: initial
    }
  })

  return renderer.create(
    <Provider store={store}>
      <IntlProvider locale="en">
        <ReduxProduct />
      </IntlProvider>
    </Provider>
  )
}

describe('component:Product', () => {
  describe('snapshots', () => {
    it('renders without crashing', () => {
      const target = setupSnapshot([])
      let tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('only shows the first five items', () => {
      const target = setupSnapshot(fixture)
      let tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('sorting', () => {
    it('places selections ahead of unselected', () => {
      const selected = [
        'Credit reporting, credit repair services, or other personal consumer reports',
        slugify('Credit reporting, credit repair services, or other personal consumer reports', 
          'Other personal consumer report'),
        'Credit card'
      ]
      const actual = mapStateToProps({ 
        query: {product: selected},
        aggs: {product: fixture}
      })
      expect(actual.options[1]).toEqual(fixture[5])
    })

    it('treats child selections as parent selections', () => {
      const selected = [
        slugify("Mortgage", 'Conventional home mortgage')
      ]
      const actual = mapStateToProps({ 
        query: {product: selected},
        aggs: {product: fixture}
      })
      expect(actual.options[0]).toEqual(fixture[1])
    })
  })
})