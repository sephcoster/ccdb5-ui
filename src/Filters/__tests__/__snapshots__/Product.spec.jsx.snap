// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:Product snapshots only shows the first five items 1`] = `
<section
  className="o-expandable aggregation"
>
  <div
    className="o-expandable_header o-expandable_target"
  >
    <h4
      className="o-expandable_header-left o-expandable_label"
    >
      Product / sub-product
    </h4>
    <span
      className="o-expandable_header-right o-expandable_link"
    >
      <button
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}
      >
        Hide
        <span
          className="cf-icon cf-icon-minus-round"
        />
      </button>
    </span>
  </div>
  <p>
    The type of product and sub-product the consumer identified in the complaint
  </p>
  <p>
    <a
      className="a-btn__link"
      href="http://files.consumerfinance.gov/f/documents/201704_cfpb_Summary_of_Product_and_Sub-product_Changes.pdf"
      target="_blank"
    >
      Recent changes to products and sub-products
    </a>
  </p>
  <div>
    <ul>
      <div
        className="aggregation-branch"
      >
        <li
          className="parent m-form-field m-form-field__checkbox body-copy"
        >
          <input
            aria-label="Credit reporting, credit repair services, or other personal consumer reports"
            checked={false}
            className="flex-fixed a-checkbox"
            id="productCreditreporting, credit repair services, or other personal consumer reports"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="toggle a-label"
            htmlFor="productCreditreporting, credit repair services, or other personal consumer reports"
          />
          <button
            className="flex-all a-btn a-btn__link"
            onClick={[Function]}
            title="Credit reporting, credit repair services, or other personal consumer reports"
          >
            <span>
              Credit reporting, credit repair services, or other personal consumer reports
            </span>
            <span
              className="cf-icon cf-icon-down"
            />
          </button>
          <span
            className="flex-fixed parent-count"
          >
            <span>
              3,277
            </span>
          </span>
        </li>
      </div>
      <div
        className="aggregation-branch"
      >
        <li
          className="parent m-form-field m-form-field__checkbox body-copy"
        >
          <input
            aria-label="Mortgage"
            checked={false}
            className="flex-fixed a-checkbox"
            id="productMortgage"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="toggle a-label"
            htmlFor="productMortgage"
          />
          <button
            className="flex-all a-btn a-btn__link"
            onClick={[Function]}
            title="Mortgage"
          >
            <span>
              Mortgage
            </span>
            <span
              className="cf-icon cf-icon-down"
            />
          </button>
          <span
            className="flex-fixed parent-count"
          >
            <span>
              2,299
            </span>
          </span>
        </li>
      </div>
      <li
        className="layout-row m-form-field m-form-field__checkbox"
      >
        <input
          aria-label="Credit reporting"
          checked={false}
          className="flex-fixed a-checkbox"
          id="productCreditreporting"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="a-label flex-all bucket-key body-copy"
          htmlFor="productCreditreporting"
        >
          Credit reporting
        </label>
        <span
          className="flex-fixed bucket-count"
        >
          <span>
            1,052
          </span>
        </span>
      </li>
      <li
        className="layout-row m-form-field m-form-field__checkbox"
      >
        <input
          aria-label="Student loan"
          checked={false}
          className="flex-fixed a-checkbox"
          id="productStudentloan"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="a-label flex-all bucket-key body-copy"
          htmlFor="productStudentloan"
        >
          Student loan
        </label>
        <span
          className="flex-fixed bucket-count"
        >
          <span>
            959
          </span>
        </span>
      </li>
      <li
        className="layout-row m-form-field m-form-field__checkbox"
      >
        <input
          aria-label="Credit card or prepaid card"
          checked={false}
          className="flex-fixed a-checkbox"
          id="productCreditcard or prepaid card"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="a-label flex-all bucket-key body-copy"
          htmlFor="productCreditcard or prepaid card"
        >
          Credit card or prepaid card
        </label>
        <span
          className="flex-fixed bucket-count"
        >
          <span>
            836
          </span>
        </span>
      </li>
    </ul>
    <div>
      <button
        className="a-btn a-btn__link more"
        onClick={[Function]}
      >
        + Show 
        1
         
        more
      </button>
    </div>
  </div>
</section>
`;

exports[`component:Product snapshots renders without crashing 1`] = `
<section
  className="o-expandable aggregation"
>
  <div
    className="o-expandable_header o-expandable_target"
  >
    <h4
      className="o-expandable_header-left o-expandable_label"
    >
      Product / sub-product
    </h4>
    <span
      className="o-expandable_header-right o-expandable_link"
    >
      <button
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}
      >
        Hide
        <span
          className="cf-icon cf-icon-minus-round"
        />
      </button>
    </span>
  </div>
  <p>
    The type of product and sub-product the consumer identified in the complaint
  </p>
  <p>
    <a
      className="a-btn__link"
      href="http://files.consumerfinance.gov/f/documents/201704_cfpb_Summary_of_Product_and_Sub-product_Changes.pdf"
      target="_blank"
    >
      Recent changes to products and sub-products
    </a>
  </p>
  <div>
    <ul />
  </div>
</section>
`;
